Parsimony/Parser.cs

```diff
@@ -0,0 +1,212 @@
+
+            public Option(OptionSpec<T> spec)
+            {
+                Spec = spec ?? throw new ArgumentNullException(nameof(spec));
+            }
+        }
+
+        private class ParserState
+        {
+            public ParseResult<T> Result { get; private set; }
+
+            public List<string> Input { get; private set; }
+
+            public ParserState(ParseResult<T> result, IEnumerable<string> input)
+            {
+                Result = result ?? throw new ArgumentNullException(nameof(result));
+                Input = input?.ToList() ?? throw new ArgumentNullException(nameof(input));
+            }
+        }
+
+        private static readonly Regex _shortNameOptionPatern =
+            new Regex("^-(?<option>[A-Za-z])(?<rest>.+)?", RegexOptions.Compiled);
+
+        private readonly List<Option> _options;
+
+        private readonly ParserConfig _configuration;
+
+        /// <summary>
+        /// Creates a new <see cref="Parser{T}"/>.
+        /// </summary>
+        /// <param name="optionSpecs">The option specifications.</param>
+        /// <param name="configuration">The parser configuration.</param>
+        /// <exception cref="ArgumentNullException">
+        /// <paramref name="optionSpecs"/> or <paramref name="configuration"/> is <c>null</c>.
+        /// </exception>
+        public Parser(IEnumerable<OptionSpec<T>> optionSpecs, ParserConfig configuration)
+        {
+            _options = (optionSpecs ?? throw new ArgumentNullException(nameof(optionSpecs)))
+                .Select(spec => new Option(spec)).ToList();
+
+            _configuration = configuration ?? throw new ArgumentNullException(nameof(configuration));
+
+            // TODO: Validate specs for duplicate long/short names.
+        }
+
+        /// <summary>
+        /// Parses a <typeparamref name="T"/> from <paramref name="input"/>.
+        /// </summary>
+        /// <param name="input">The tokens to parse.</param>
+        /// <returns>The <see cref="ParseResult{T}"/>.</returns>
+        /// <exception cref="ArgumentNullException">
+        /// <paramref name="input"/> is <c>null</c>.
+        /// </exception>
+        public ParseResult<T> Parse(IEnumerable<string> input)
+        {
+            var tokens = (input ?? throw new ArgumentNullException(nameof(input))).ToList();
+
+            var options = new T();
+            var arguments = new List<string>(tokens.Count);
+            var result = new ParseResult<T>(options, arguments);
+            return Parse(result, tokens);
+        }
+
+        /// <summary>
+        /// Parses a <typeparamref name="T"/> from <paramref name="input"/>.
+        /// </summary>
+        /// <param name="input">The tokens to parse.</param>
+        /// <returns>The <see cref="ParseResult{T}"/>.</returns>
+        /// <exception cref="ArgumentNullException">
+        /// <paramref name="input"/> is <c>null</c>.
+        /// </exception>
+        public ParseResult<T> Parse(params string[] input)
```

Comment:
why?

Reply:
just so I could write `p.Parse("foo", "bar", "baz")` instead of `p.Parse(new [] { "foo", "bar", "baz" })` in the tests. I was thinking consumes might touch this interface in testing too. I suppose I could move this to an ext in the testing lib and if someone else wants it and can't figure out to do the same then too bad for them